bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/obsComp/TargetSelection.java

```diff
@@ -34,110 +36,144 @@ private TargetSelection() {}
         Index(int value) { this.value = value; }
     }
 
+    // Used internally as a placeholder for None.
     private static final Index NO_SELECTION = new Index(-1);
 
     private static boolean isValid(final ISPNode node) {
-        return (node instanceof ISPObsComponent) && ((ISPObsComponent) node).getType() == SPComponentType.TELESCOPE_TARGETENV;
+        return (node instanceof ISPObsComponent) &&
+                ((ISPObsComponent) node).getType() == SPComponentType.TELESCOPE_TARGETENV;
     }
 
-    public static int getIndex(final ISPNode node) {
-        if (!isValid(node)) return NO_SELECTION.value;
-        final Object cd = node.getTransientClientData(KEY);
-        return ((cd == null) ? NO_SELECTION : (Index) cd).value;
+    public static Option<Integer> getIndex(final ISPNode node) {
+        if (!isValid(node)) return None.instance();
+        return ImOption.apply((Index) node.getTransientClientData(KEY))
+                .map(i -> i.value).filter(i -> i != NO_SELECTION.value);
     }
 
-    public static void setIndex(final ISPNode node, final int index) {
-        if (isValid(node)) node.putTransientClientData(KEY, new Index(index));
+    private static void setIndex(final ISPNode node, final int index) {
+        setIndex(node, index == -1 ? None.instance() : new Some<>(index));
     }
 
+    public static void setIndex(final ISPNode node, final Option<Integer> index) {
+        if (isValid(node)) node.putTransientClientData(KEY, index.map(Index::new).getOrElse(NO_SELECTION));
+    }
+
+    /**
+     * Marks a selection object.
+     */
     private static final class Selection {
-        final GuideGroup guideGroup;
-        final SPTarget target;
+        final int index;
+        final Option<GuideGroup> guideGroup;
+        final Option<SPTarget> target;
 
-        private Selection(final GuideGroup grp, final SPTarget target) {
+        private Selection(final int index, final Option<GuideGroup> grp, final Option<SPTarget> target) {
```

Comment:
I think it's great that you're cleaning this up and better representing what we're trying to model.  Could/should we take it a step further and make `Selection` an interface with three implementations: `NormalTargetSelection` which has an `SPTarget`, `GroupSelection` which has a `GuideGroup`, and `GuideStarSelection` which has both?  I just noticed that we can still construct this with (int, None, None).